### PR TITLE
feat(CI): Add a new task to automatically assign a team label on a dependabot PR

### PR DESCRIPTION
**Pull request by chouetz**
### What does this PR do?
- Add a new task to automatically assign a team label on a dependabot PR
- Add an associated workflow to assign at PR creation

### Motivation
Remove manual actions

### Possible Drawbacks / Trade-offs
Assignment is based on the occurrence of dependency found in repository files. This heuristic is basic and prevents the PR from being unassigned. It does not prevent a manual redirection.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: .github/workflows/assign-dependabot.yml
Comment (chouquette) [inline .github/workflows/assign-dependabot.yml:4]: Should this specify `types: [created]` to avoid further runs once the initial dispatch is complete?

File: tasks/assign_pr.py
Comment (chouquette) [inline tasks/assign_pr.py:62]: I absolutely don't mind switching to the `gh` cli tool instead of using their API, but we might want to pick one and stick with it

File: tasks/assign_pr.py
Comment (chouquette) [inline tasks/assign_pr.py:47]: Is this expected to happen? I'd expect dependabot's PR to be properly unicode encoded :worried:

File: .github/workflows/assign-dependabot.yml
Comment (chouquette) [inline .github/workflows/assign-dependabot.yml:13]: Out of curiosity, why not using `${{ github.event.number }}` directly? It's only used once

File: tasks/assign_pr.py
Comment (chouquette) [inline tasks/assign_pr.py:20]: You could probably pass the PR title directly from the action to save a roundtrip

File: tasks/assign_pr.py
Comment (chouquette) [inline tasks/assign_pr.py:35]: Shouldn't you only check for `doc` in the beginning of the path? I suppose the intent is to map any change to the `docs` folder but that will likely match for quite a few more patterns as is

File: tasks/assign_pr.py
Comment (chouquette) [inline tasks/assign_pr.py:42]: what are the 2 characters you're removing from the file_path?

Comment (pgimalac) [inline tasks/assign_pr.py:42]: Why remove the first two characters of the path ? Is it `./` ?
If so you might want to use `os.path.normpath()` to normalize the path

File: tasks/assign_pr.py
Comment (pgimalac) [inline tasks/assign_pr.py:62]: You might want to request review from those teams too ?
I know I usually don't look at PRs with my team's label (except just before freeze) but I do look at all PRs my team is assigned on

File: tasks/assign_pr.py
Comment (pgimalac) [inline tasks/assign_pr.py:76]: This doesn't seem very reliable, and is already missing most prefixes (see the root `go.mod` for many examples, eg `cloud.google.com`, `code.cloudfoundry.org`, `contrib.go.opencensus.io`, etc)

File: tasks/assign_pr.py
Comment (pgimalac) [inline tasks/assign_pr.py:58]: I'm not sure assigning the team who owns the most files using the dependency is the best way, I would expect all teams using the dependency to be assigned...

Possibly with an override list, eg. `github.com/stretchr/testify` for agent-platform, `golang.org/x/*` and `google.golang.org` for agent-shared-components, etc

File: tasks/assign_pr.py
Comment (pgimalac) [inline tasks/assign_pr.py:40]: I think a better way to check if it's a go module is to check if the PR edits a `go.mod` file.
You should probably add a check that the file is a `go` or `go.mod` file for this `if` block

File: tasks/assign_pr.py
Comment (pgimalac) [inline tasks/assign_pr.py:1]: I think it would be nicer to use the labels `dependencies-go`, `dependencies-python`, etc to determine which kind of dependency is updated, and then have an implementation for each language to only read the needed files and have a more precise matching on when the dependency is used

File: tasks/assign_pr.py
Comment (pgimalac) [inline tasks/assign_pr.py:30]: If you don't want a super specific implementation you might want to just use a `grep` command to list files containing a given string, it would simplify this whole function
https://stackoverflow.com/a/16957078/10556221

**Review by pgimalac — APPROVED** [review]
Wrote a few comments which came to my mind when reading the code, I understand this needs to handle different languages so having a single perfect implementation is impossible, but I'm not sure this is the best way to do it 😄 

EDIT: in particular I think this job only makes sense for the root `go.mod` file, so the code could be way more specialized for that (only Go, only root module)

File: .github/workflows/assign-dependabot.yml
Comment (pgimalac) [reply .github/workflows/assign-dependabot.yml:4]: I had doubts about that too but according to the [doc](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request):
> By default, a workflow only runs when a pull_request event's activity type is opened, synchronize, or reopened

Comment (pgimalac) [reply .github/workflows/assign-dependabot.yml:4]: Edit: actually `synchronize` is basically a push event sooo yeah fair point 😄

File: tasks/assign_pr.py
Comment (pgimalac) [reply tasks/assign_pr.py:58]: `label for label in team_labels if responsible in label` is an iterator, `next` tries to take the next (first in that case) one

File: .github/workflows/assign-dependabot.yml
Comment (chouetz) [reply .github/workflows/assign-dependabot.yml:4]: hm yes I could probably limit to `opened, reopened` because it seems dependabot can continue pushing commits to an opened branch (even if the label should remain the same so "should" be harmless

File: tasks/assign_pr.py
Comment (chouetz) [reply tasks/assign_pr.py:62]: it was just easier for such actions :/

File: tasks/assign_pr.py
Comment (chouetz) [reply tasks/assign_pr.py:47]: It's because I `walk` through the whole repository which contains binary files

File: tasks/assign_pr.py
Comment (chouetz) [reply tasks/assign_pr.py:20]: Let's add the title as input, ok

File: tasks/assign_pr.py
Comment (chouetz) [reply tasks/assign_pr.py:42]: because as I put `"."` as base directory, all paths are `./...` and it prevents matching in the codeowners file

File: tasks/assign_pr.py
Comment (chouquette) [reply tasks/assign_pr.py:62]: Oh don't get me wrong I'm not saying you must use the API here because we use it elsewhere! But there are probably other places where using the `gh` cli would be simpler.
This is definitely not a blocking comment

**Comment by chouetz** [comment]
> Wrote a few comments which came to my mind when reading the code, I understand this needs to handle different languages so having a single perfect implementation is impossible, but I'm not sure this is the best way to do it 😄
> 
> EDIT: in particular I think this job only makes sense for the root `go.mod` file, so the code could be way more specialized for that (only Go, only root module)

dependabot updates all version of various dependencies. As a consequence it works also for github actions updates for instance

File: tasks/assign_pr.py
Comment (spencergilbert) [reply tasks/assign_pr.py:62]: I'd support an effort to move to `gh`

File: tasks/assign_pr.py
Comment (chouetz) [reply tasks/assign_pr.py:1]: I think I prefer to stick with the ownership where the dependency is effectively imported, and I understand your point. Assigning to the owner of the `go.mod` would mean we should not have a different ownership in the module, and I don't know if this is always true, wdyt?

File: tasks/assign_pr.py
Comment (chouetz) [reply tasks/assign_pr.py:42]: Indeed I can use this it's better

File: tasks/assign_pr.py
Comment (chouetz) [reply tasks/assign_pr.py:62]: Currently the role said we just have to assign a label to the correct team. We might want to clarify the process and responsabilities here as I've seen we have "just" 200+ dependabot PR opened which is a mess

File: tasks/assign_pr.py
Comment (chouetz) [reply tasks/assign_pr.py:58]: * About one team versus several ones I agree we can discuss. Again with roles and responsibilities to be clarified so that the assigned one must validate and/or approve or add to MQ
* Regarding the override list, yes it's a valid improvement. Current implementation was approximative on purpose, should not prevent manual redirection as said in description

File: tasks/assign_pr.py
Comment (chouetz) [reply tasks/assign_pr.py:40]: CF comment above. Either I consider only the `go.mod` file and I think it's not correct in terms of ownership, or I consider everything and so a pattern in the dependency looked like an acceptable approximation (knowing there is another match in the else)

File: tasks/assign_pr.py
Comment (chouetz) [reply tasks/assign_pr.py:30]: I agree I use calls to `gh`, meanwhile I'm not a super fan of python code calling too many external tools..

File: tasks/assign_pr.py
Comment (chouetz) [reply tasks/assign_pr.py:35]: I will match `docs/` instead (it's not necessarily a `startswith` as you have docs in test as well)

**Comment by pgimalac** [comment]
> dependabot updates all version of various dependencies. As a consequence it works also for github actions updates for instance

@chouetz yes but workflows are owned by a specific team, so the PR will already be owned by that team.
The root go mod file is the only source of dependencies which doesn't have an owner and whose scope is broader in terms of teams.

File: tasks/assign_pr.py
Comment (pgimalac) [reply tasks/assign_pr.py:76]: Fair, but in that case you could just handle it in a single way no matter if it's a Go dependency or not

File: tasks/assign_pr.py
Comment (pgimalac) [reply tasks/assign_pr.py:1]: I gave it a quick look and it seems to me the single module for which it's not the case is `test/new-e2e`.
All other modules (ignoring the root one) have a narrow scope and a specific owner.

File: tasks/assign_pr.py
Comment (pgimalac) [reply tasks/assign_pr.py:30]: I think calling `grep` is pretty trivial, it doesn't have side effects or anything, it's just a very convenient tool
But up to you :)

File: tasks/assign_pr.py
Comment (chouetz) [reply tasks/assign_pr.py:76]: Added few of them from the root `go.mod`

File: tasks/assign_pr.py
Comment (pgimalac) [reply tasks/assign_pr.py:62]: Real question: does `gh` provide everything the API does ?
I had some fun writing scripts to get some (relatively) complex info about PRs recently and I'm curious if I could have just used `gh` instead :smile:
FYI I used https://github.com/PyGithub/PyGithub which provides super convenient wrappers around the API, everything is typed, fetched lazily, it handles pagination, etc

**Comment by chouetz** [comment]
> > dependabot updates all version of various dependencies. As a consequence it works also for github actions updates for instance
> 
> @chouetz yes but workflows are owned by a specific team, so the PR will already be owned by that team. The root go mod file is the only source of dependencies which doesn't have an owner and whose scope is broader in terms of teams.

Yes the objective would also to rely only on CODEOWNERS as ownership definition. You can see in this [PR](https://github.com/DataDog/datadog-agent/pull/21613) dependabot set `team/agent-platform`, and my script would rather infer `team/serverless`.
I think we could be more accurate if we have file ownership on a single place (I'm not confident about our capacity to maintain accurate ownerships on dependabot.yml and in CODEOWNERS ;) )

**Comment by pgimalac** [comment]
Oh ok I assumed for this kind of situation we just didn't set a label in the dependabot config and let the real code owner set its own team label, but I think I understand better what the goal of the PR is now

**Pull request closed without merging**
